muon/_core/preproc.py

```diff
@@ -155,9 +155,25 @@ def _l2norm(
     adata: AnnData, rep: Optional[Union[Iterable[str], str]] = None, n_pcs: Optional[int] = 0
 ):
     X = _choose_representation(adata, rep, n_pcs)
-    norm = X / np.linalg.norm(X, ord=2, axis=1, keepdims=True)
-    norm[~np.isfinite(norm)] = 0
+    sparse_X = issparse(X)
+    if sparse_X:
+        X_norm = linalg.norm(X, ord=2, axis=1)
+        norm = X / np.expand_dims(X_norm, axis=1)
+        if issparse(norm):
+            # find nan and infinite values and construct matrix without
+            i, j, val = find(norm)
+            isfin = np.isfinite(val)
+            i = i[isfin]
+            j = j[isfin]
+            val = val[isfin]
+            norm = csr_matrix((val, (i, j)), shape=X.shape)
+    else:
+        norm = X / np.linalg.norm(X, ord=2, axis=1, keepdims=True)
+        norm[~np.isfinite(norm)] = 0
     X.astype(norm.dtype, copy=False)
+    if sparse_X and not issparse(norm):
+        norm[~np.isfinite(norm)] = 0
+        X = X.toarray()
     X[:] = norm
```

Comment:
I'm not sure at the moment, but can you test if this does what we expect it to do for sparse matrices? I.e. it doesn't change the sparsity structure? Otherwise it would perhaps make sense to do something like
```
if sparse_X and (isspmatrix_csc(X) or isspmatrix_csr(X) or issspmatrix_coo(X)):
    X.data[:] = norm.data[:]
else:
    X[:] = norm
```
(I don't know enough about the other sparse matrix types off the top of my head, but this should cover most usecases)

Reply:
I had a weird case under Python 3.8 when `norm` became dense, and then `X[:] = norm` crashed my kernel. That's why I wanted to catch that error.

Reply:
I'll test your suggestion.

Reply:
I tested your suggestion and included it in the code with minimal adaptations (commit https://github.com/scverse/muon/pull/114/commits/cd3f942212353a4b1ed52902e587174d3858ec56).

Reply:
Following up on the part where `norm` became dense and `X` was still sparse. I added a clause to convert `norm` into a `csr_matrix` in that case.